File: src/main/kotlin/org/kite9/diagram/common/algorithms/fg/LinearArc.java
package org.kite9.diagram.common.algorithms.fg;


/**
 * A linear arc has a constant cost for each item of flow pushed into it.
 * It has upper and lower bounds on the amount of flow that can travel through it.
 * @author Rob Moffat
 *
 */
public class LinearArc extends AbstractArc {

	protected int cost;
	
	protected int lowerBound;
	
	protected int capacity;

	public LinearArc(int cost, int capacity, int lowerBound, Node from, Node to, String label) {
		super();
		this.cost = cost;
		this.capacity = capacity;
		this.from = from;
		this.to = to;
		this.lowerBound = lowerBound;
		this.label = label;
		from.getArcs().add(this);
		to.getArcs().add(this);
	}

	public int getIncrementalCost(int flow) {
		return flow * cost;
	}

	public boolean hasCapacity(boolean reversed) {
		if (!reversed) {
			return flow < capacity;
		} else {
			return flow > lowerBound;
		}
	}
	
}

File: a/src/myMain/kotlin/org/kite9/diagram/common/algorithms/fg/LinearArc.kt
package org.kite9.diagram.common.algorithms.fg

import org.kite9.diagram.common.algorithms.fg.AbstractArc

/**
 * A linear arc has a constant cost for each item of flow pushed into it.
 * It has upper and lower bounds on the amount of flow that can travel through it.
 * @author Rob Moffat
 */
class LinearArc(
    protected var cost: Int,
    protected var capacity: Int,
    protected var lowerBound: Int,
    from: Node,
    to: Node,
    label: String
) : AbstractArc(from, to, label, 0) {

    override fun getIncrementalCost(flow: Int): Int {
        return flow * cost
    }

    override fun hasCapacity(reversed: Boolean): Boolean {
        return if (!reversed) {
            flow < capacity
        } else {
            flow > lowerBound
        }
    }

}